### PR TITLE
*: add Docker build environment

### PR DESCRIPTION
**Pull request by mrd0ll4r**
I'm not 100% sure if it's worth executing the builder stage in CI every time.
I guess it's fine -- the more things we test continuously, the better.
At some point in the future we could make these available somehow...

This depends on #23

### CODE REVIEW

File: Dockerfile.builder [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+RUN go build -v -o /usr/local/bin/ipfs/ipfs-v0.12.0-docker ./cmd/ipfs
+
+# Compile metric export plugin
+WORKDIR /usr/src/ipfs/metric-export-plugin
+# Get and compile dependencies first, since we can cache those.
+COPY go.mod go.sum ./
+RUN go mod download && go mod verify
+
+# Then copy sources and compile project
+COPY . .
```
Comment (scriptkitty) [inline Dockerfile.builder:26]: This feels a bit wonky but I guess just copying all the things does no real harm either

File: README.md [hunk 2]
```diff
@@ -26,6 +26,18 @@ Manually, building with `go build -buildmode=plugin -o mexport.so` should also w
...
+
+You can build this project together with a matching go-ipfs executable within Docker.
+This is nice, because you get reproducible, matching binaries, compiled with Go 1.16 on Debian bullseye.
+Building on bullseye gives us a libc version which is a bit older.
+This gives us compatibility with slightly older systems (e.g. Ubuntu LTS releases), at no loss of functionality.
+
+The [builder Dockerfile](./Dockerfile.builder) implements a builder stage.
+The resulting binaries are placed in `/usr/local/bin/ipfs/` inside the image.
+
+The [build-in-docker.sh](./build-in-docker.sh) script executes the builder and copies the produced binaries to the `out/` directory.
```
Comment (scriptkitty) [inline README.md:39]: To the out dir of the project not the image, right?

File: Dockerfile.builder [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+# We build go-ipfs first, because, since we're building off a defined tag, the sources don't change and this can be
+# cached.
+WORKDIR /usr/src/ipfs/go-ipfs
+RUN git clone https://github.com/ipfs/go-ipfs.git . && git checkout v0.12.0
+RUN go build -v -o /usr/local/bin/ipfs/ipfs-v0.12.0-docker ./cmd/ipfs
+
+# Compile metric export plugin
+WORKDIR /usr/src/ipfs/metric-export-plugin
+# Get and compile dependencies first, since we can cache those.
+COPY go.mod go.sum ./
```
Comment (scriptkitty) [inline Dockerfile.builder:22]: Not a big deal at the moment but just for completntess sake: This means that a new image is created everytime the code changes and the CI is executed, correct?
Maybe just having go-ipfs with a fixed tag inside the container and simply mounting the plugin source would be more efficient?

**Review by scriptkitty — APPROVED** [review]

**Comment by mrd0ll4r** [comment]
Some dependency of the format checker is broken, I'll try to re-run the CI and merge in a few days...

**Comment by mrd0ll4r** [comment]
Ah, turns out this is because we use Go 1.16, and the latest version of `mvdan.cc/gofumpt`, which is `v0.3.1` at the moment, uses a new syntax for go:build annotations, which the old compiler does not understand. This is popping up now because Go 1.18 was release, which means Go 1.16 is no longer supported and gofumpt released the new version. Both 1.17 and 1.18 understand the new annotations.

We cannot build with Go 1.17 due to the linker bugs with plugins.
We cannot yet build with Go 1.18 due to an [outdated dependency on go-quic](https://github.com/ipfs/go-ipfs/issues/8819).

As a temporary workaround I'll fix the version of `gofumpt` to `v0.3.0`, which still compiles with Go 1.16.

**Pull request merged**